RecoParticleFlow/PFProducer/plugins/PFProducer.cc

```diff
@@ -333,3 +309,233 @@ void PFProducer::produce(Event& iEvent, const EventSetup& iSetup) {
     iEvent.put(muAlgo.transferAddedMuonCandidates(), "AddedMuonsAndHadrons");
   }
 }
+
+void PFProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  // Verbosity and debug flags
+  desc.addUntracked<bool>("verbose", false);
+  desc.addUntracked<bool>("debug", false);
+
+  // PF Blocks label
+  desc.add<edm::InputTag>("blocks", edm::InputTag("particleFlowBlock"));
+
+  // reco::muons label and Post Muon cleaning
+  desc.add<edm::InputTag>("muons", edm::InputTag("muons1stStep"));
+  desc.add<bool>("postMuonCleaning", true);
+
+  // Vertices label
+  desc.add<edm::InputTag>("vertexCollection", edm::InputTag("offlinePrimaryVertices"));
+  desc.add<bool>("useVerticesForNeutral", true);
+
+  // Use HO clusters in PF hadron reconstruction
+  desc.add<bool>("useHO", true);
+
+  // EGamma-related
+  desc.add<edm::InputTag>("PFEGammaCandidates", edm::InputTag("particleFlowEGamma"));
+  desc.add<edm::InputTag>("GedElectronValueMap", edm::InputTag("gedGsfElectronsTmp"));
+  desc.add<edm::InputTag>("GedPhotonValueMap", edm::InputTag("gedPhotonsTmp", "valMapPFEgammaCandToPhoton"));
+
+  desc.add<bool>("useEGammaElectrons", true);
+  desc.add<edm::InputTag>("egammaElectrons", edm::InputTag("mvaElectrons"));
+
+  desc.add<bool>("useEGammaFilters", true);
+  desc.add<bool>("useProtectionsForJetMET", true);  // Propagated to PFEGammaFilters
+
+  // For PFEGammaFilters
+  {
+    edm::ParameterSetDescription psd_PFEgammaFilters;
```

Comment:
it looks like this PSet  description fully belongs to the PFEgammaFilters (not used anywhere else by the PFProducer). This is a pattern usually implies that the defaults better be defined by ```PFEgammaFilters``` 

A suggested approach for this for a general case is https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#A_Plugin_Module_Using_Another_He
Here, since the factory pattern is not used for the helper class, a simpler solution is to define in ```PFEgammaFilters```
```c++
static void
fillPSetDescription(edm::ParameterSetDescription& iDesc) {
    iDesc.add<double>("electron_iso_pt", 10.0);
 ...
}
```

Reply:
I have to admit I am still not sure how to do this and need help.
We don't really use "factory", so I am not sure how this recipe will fit it into this use case.

Some specific questions are:
+ PFEGammaFilters may be considered as a helper, but it's not a plugin module right now. We need to make it as a plugin?
+ Do we need to introduce a new "factory" that creates PFEGammaFilters as a new plugin?
I think I need a little more guidance to get this implemented.

Reply:
Maybe a little more in detail, by looking at:
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#A_Plugin_Module_Using_Another_He
I probably understand steps 1,2,3, but steps 4,5,6 are unclear (which files these lines need to go in?) to me right now.

Reply:
the link to the twiki was given for a more complete reference for a general case.
For this specific case, please follow the suggested code snippets:

add in PFEgammaFilters.h,cc: (as already posted above)
```c++
static void PFEgammaFilters::fillPSetDescription(edm::ParameterSetDescription& iDesc) {
    iDesc.add<double>("electron_iso_pt", 10.0);
 ...
}
```
then in this file replace L349-L409 with 
```c++
PFEgammaFilters::fillPSetDescription(psd_PFEgammaFilters);
```